### PR TITLE
[IoT] Fix inconsistency in MqttManager.disconnect()

### PR DESCRIPTION
**Pull request by kvasukib**
*Issue #, if available:*

#292 

*Description of changes:*

* **AWS IoT**
  * Fix an issue where the MqttClient is stuck disconnecting. See [issue #292](https://github.com/aws-amplify/aws-sdk-android/issues/292) for details.
  * `MqttManager.disconnect()` now will do a disconnect forcibly irrespective of the state of the MQTT connection.
     * `MqttManager.disconnect()` returns true when the disconnect is successful and false otherwise.
     * `MqttManager.disconnect()` will disconnect the connection even if the client is reconnecting.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java [hunk 1]
```diff
@@ -997,32 +1002,61 @@ public void onFailure(IMqttToken asyncActionToken, Throwable e) {
...
-     * Disconnect from a mqtt client (close current MQTT session).
-     *
-     * @return true if disconnect finished with success.
-     */
     public boolean disconnect() {
         userDisconnect = true;
-        reset();
+
+        if (null != mqttClient) {
+            if (mqttClient.isConnected()) {
```
Comment (desokroshan) [inline aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java]: Can we combine the two if-conditions?

File: aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java [hunk 2]
```diff
@@ -997,32 +1002,61 @@ public void onFailure(IMqttToken asyncActionToken, Throwable e) {
...
+                // Disconnect with 0 seconds quiesce timeout. On successful disconnect from Paho,
+                // clear the topic listeners, set the connection state to Disconnected. Invoke the
+                // connection callback if registered and invoke the disconnect callback.
+                // On any error in disconnect from Paho, invoke the callback with the appropriate error.
+                mqttClient.disconnect(0,null, new IMqttActionListener() {
+                    @Override
+                    public void onSuccess(IMqttToken asyncActionToken) {
+                        topicListeners.clear();
+                        connectionState = MqttManagerConnectionState.Disconnected;
+                        userConnectionCallback();
```
Comment (desokroshan) [inline aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java]: These statements seem to be repeated.

File: aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java [hunk 3]
```diff
@@ -46,10 +47,14 @@
...
 import java.security.NoSuchProviderException;
 import java.security.UnrecoverableKeyException;
+import java.util.ArrayList;
 import java.util.HashMap;
+import java.util.List;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
```
Comment (desokroshan) [inline aws-android-sdk-iot/src/main/java/com/amazonaws/mobileconnectors/iot/AWSIotMqttManager.java]: All these imports except the Callback seems to be unused

**Comment by kvasukib** [comment]
Closing this PR as the new disconnect method does not offer any benefit compared to the old one. And #292 has been updated with the root cause and findings.

**Pull request closed without merging**